### PR TITLE
Implement Connection Lifetime

### LINKED ISSUE #212: Implement Connection Lifetime
[Official description](https://dev.mysql.com/doc/connector-net/en/connector-net-connection-options.html):

> When a connection is returned to the pool, its creation time is compared with the current time, and the connection is destroyed if that time span (in seconds) exceeds the value specified by `Connection Lifetime`. This is useful in clustered configurations to force load balancing between a running server and a server just brought online. A value of zero (0) causes pooled connections to have the maximum connection timeout.

This appears to enforce a maximum lifetime on any individual connection that is created. It's not clear to me (from the documentation) if some kind of watchdog is proactively cleaning up old connections in the pool, or if this only applies strictly to when a connection is returned.

### PR DESCRIPTION
**Pull request by caleblloyd**
- Fixes #212

### CODE REVIEW

File: tests/SideBySide/ConnectionPool.cs [hunk 1]
```diff
@@ -160,6 +160,37 @@ public void ResetConnection(object connectionReset, int poolSize, long expected)
...
+			int serverThread;
+
+			using (var connection = new MySqlConnection(csb.ConnectionString))
+			{
+				await connection.OpenAsync();
+				serverThread = connection.ServerThread;
+				await Task.Delay(TimeSpan.FromSeconds(1));
+			}
+
+			using (var connection = new MySqlConnection(csb.ConnectionString))
```
Comment (caleblloyd) [inline tests/SideBySide/ConnectionPool.cs]: Baseline fails here in AppVeyor: https://ci.appveyor.com/project/mysqlnet/mysqlconnector/build/1.0.389#L423

File: src/MySqlConnector/MySqlClient/ConnectionPool.cs [hunk 2]
```diff
@@ -53,11 +53,26 @@ internal sealed class ConnectionPool
...
+		private bool SessionIsHealthy(MySqlSession session)
+		{
+			if (!session.IsConnected)
+				return false;
+			if (session.PoolGeneration != m_generation)
+				return false;
+			if (session.DatabaseOverride != null)
+				return false;
+			if (m_connectionSettings.ConnectionLifeTime > 0
+			    && DateTime.UtcNow - session.CreatedUtc > TimeSpan.FromSeconds(m_connectionSettings.ConnectionLifeTime))
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/ConnectionPool.cs]: Nitpick: I might personally write `(DateTime.UtcNow - session.CreatedUtc).TotalSeconds >= m_connectionSettings.ConnectionLifeTime` (slightly fewer keystrokes, avoids the creation of a temporary) but it's hardly worth changing.

**Comment by caleblloyd** [comment]
> Baseline tests are failing; is this not actually implemented correctly in MySql.Data?

They pass for me locally but fail on AppVeyor.  I was wondering if maybe it's something to do with AppVeyor's speed, it takes me 30 seconds to run all baseline tests locally but AppVeyor is 1-2 mins.

**Comment by bgrainger** [comment]
Tests are working locally for me, too.

**Comment by caleblloyd** [comment]
I'll add some debug output for appveyor

**Comment by caleblloyd** [comment]
```
ConnectionLifeTime 30 Open 1 at 4/6/2017 7:32:49 PM
ConnectionLifeTime 30 Open 2 at 4/6/2017 7:33:06 PM
```

That's not 3 seconds!  I guess this is what Sync-Over-Async gets you.

**Pull request merged**